File: packages/sdk/rust/packages/core/src/http_client.rs
//! EdgeBase Rust SDK — Internal HTTP client (#133: set_context / X-EdgeBase-Context removed)

use crate::Error;
use reqwest::{Client, Method};
use serde_json::Value;
use std::time::Duration;

pub struct HttpClient {
    client: Client,
    base_url: String,
    service_key: String,
    #[cfg_attr(not(test), allow(dead_code))]
    timeout_ms: Option<u64>,
}

fn parse_timeout_ms(raw: Option<&str>) -> Option<u64> {
    raw.and_then(|value| value.trim().parse::<u64>().ok())
        .filter(|value| *value > 0)
}

impl HttpClient {
    pub fn new(base_url: &str, service_key: &str) -> Result<Self, Error> {
        let url = base_url.trim_end_matches('/').to_string();
        let timeout_ms = parse_timeout_ms(std::env::var("EDGEBASE_HTTP_TIMEOUT_MS").ok().as_deref());
        let mut builder = Client::builder();
        if let Some(timeout_ms) = timeout_ms {
            builder = builder.timeout(Duration::from_millis(timeout_ms));
        }
        Ok(Self {
            client: builder.build()?,
            base_url: url,
            service_key: service_key.to_string(),
            timeout_ms,
        })
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    #[cfg_attr(not(test), allow(dead_code))]
    pub(crate) fn timeout_ms(&self) -> Option<u64> {
        self.timeout_ms
    }

    #[cfg_attr(not(test), allow(dead_code))]
    pub(crate) fn parse_timeout_ms_for_tests(raw: Option<&str>) -> Option<u64> {
        parse_timeout_ms(raw)
    }

    fn build_request(&self, method: Method, path: &str) -> reqwest::RequestBuilder {
        let url = format!("{}{}", self.base_url, path);
        let mut req = self.client.request(method, &url);
        // Token refresh failed — proceed as unauthenticated
        if let Ok(key) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| self.service_key.clone())) {
            if !key.is_empty() {
                req = req.header("X-EdgeBase-Service-Key", &key);
                req = req.header("Authorization", format!("Bearer {}", key));
            }
        }
        req
    }

    async fn send(&self, req: reqwest::RequestBuilder) -> Result<Value, Error> {
        let resp = req.send().await?;
        let status = resp.status();
        let text = resp.text().await?;
        if !status.is_success() {
            let msg = serde_json::from_str::<Value>(&text)
                .ok()
                .and_then(|v| {
                    v.get("error")
                        .or_else(|| v.get("message"))
                        .and_then(|m| m.as_str())
                        .map(|s| s.to_string())
                })
                .unwrap_or_else(|| text.clone());
            return Err(Error::Api {
                status: status.as_u16(),
                message: msg,
            });
        }
        if text.is_empty() {
            return Ok(Value::Null);
        }
        Ok(serde_json::from_str(&text)?)
    }

    /// Send request with 429 retry and transport retry. Rebuilds request on each attempt.
    async fn send_with_retry(
        &self,
        method: Method,
        path: &str,
        body: Option<&Value>,
        query: Option<&std::collections::HashMap<String, String>>,
    ) -> Result<Value, Error> {
        let max_retries: usize = 3;
        for attempt in 0..=max_retries {
            let mut req = self.build_request(method.clone(), path);
            if let Some(b) = body {
                req = req.json(b);
            }
            if let Some(q) = query {
                req = req.query(q);
            }

            match req.send().await {
                Ok(resp) => {
                    let status = resp.status();
                    // 429 retry with Retry-After header
                    if status.as_u16() == 429 && attempt < max_retries {
                        let retry_after = resp
                            .headers()
                            .get("retry-after")
                            .and_then(|v| v.to_str().ok())
                            .and_then(|v| v.parse::<u64>().ok());
                        let base_ms = retry_after.map(|s| s * 1000).unwrap_or(1000 * (1u64 << attempt));
                        // Jitter: 0–25% of base delay (use simple pseudo-random from time nanos)
                        let nanos = std::time::SystemTime::now()
                            .duration_since(std::time::UNIX_EPOCH)
                            .map(|d| d.subsec_nanos())
                            .unwrap_or(0);
                        let jitter = (base_ms as f64 * 0.25 * (nanos % 1_000_000) as f64 / 1_000_000.0) as u64;
                        let delay = std::cmp::min(base_ms + jitter, 10000);
                        tokio::time::sleep(Duration::from_millis(delay)).await;
                        continue;
                    }
                    let text = resp.text().await?;
                    if !status.is_success() {
                        let msg = serde_json::from_str::<Value>(&text)
                            .ok()
                            .and_then(|v| {
                                v.get("error")
                                    .or_else(|| v.get("message"))
                                    .and_then(|m| m.as_str())
                                    .map(|s| s.to_string())
                            })
                            .unwrap_or_else(|| text.clone());
                        return Err(Error::Api {
                            status: status.as_u16(),
                            message: msg,
                        });
                    }
                    if text.is_empty() {
                        return Ok(Value::Null);
                    }
                    return Ok(serde_json::from_str(&text)?);
                }
                Err(e) => {
                    if attempt < 2 {
                        tokio::time::sleep(Duration::from_millis(50 * (attempt as u64 + 1))).await;
                        continue;
                    }
                    return Err(e.into());
                }
            }
        }
        Err(Error::Api {
            status: 0,
            message: "Request failed after retries".to_string(),
        })
    }

    pub async fn get(&self, path: &str) -> Result<Value, Error> {
        self.send_with_retry(Method::GET, path, None, None).await
    }

    pub async fn get_with_query(&self, path: &str, query: &std::collections::HashMap<String, String>) -> Result<Value, Error> {
        self.send_with_retry(Method::GET, path, None, Some(query)).await
    }

    pub async fn post(&self, path: &str, body: &Value) -> Result<Value, Error> {
        self.send_with_retry(Method::POST, path, Some(body), None).await
    }

    pub async fn post_with_query(&self, path: &str, body: &Value, query: &std::collections::HashMap<String, String>) -> Result<Value, Error> {
        self.send_with_retry(Method::POST, path, Some(body), Some(query)).await
    }

    pub async fn patch(&self, path: &str, body: &Value) -> Result<Value, Error> {
        self.send_with_retry(Method::PATCH, path, Some(body), None).await
    }

    pub async fn delete(&self, path: &str) -> Result<Value, Error> {
        self.send_with_retry(Method::DELETE, path, None, None).await
    }

    pub async fn delete_with_body(&self, path: &str, body: &Value) -> Result<Value, Error> {
        self.send_with_retry(Method::DELETE, path, Some(body), None).await
    }

    /// HEAD request — returns `true` if the resource exists (2xx status).
    pub async fn head(&self, path: &str) -> Result<bool, Error> {
        let req = self.build_request(Method::HEAD, path);
        let resp = req.send().await?;
        Ok(resp.status().is_success())
    }

    pub async fn put(&self, path: &str, body: &Value) -> Result<Value, Error> {
        self.send_with_retry(Method::PUT, path, Some(body), None).await
    }

    pub async fn put_with_query(&self, path: &str, body: &Value, query: &std::collections::HashMap<String, String>) -> Result<Value, Error> {
        self.send_with_retry(Method::PUT, path, Some(body), Some(query)).await
    }

    /// Multipart file upload.
    pub async fn upload_multipart(
        &self, path: &str, key: &str, data: Vec<u8>, content_type: &str,
    ) -> Result<Value, Error> {
        use reqwest::multipart::{Form, Part};
        let part = Part::bytes(data)
            .file_name(key.to_string())
            .mime_str(content_type)
            .map_err(|e| Error::Url(e.to_string()))?;
        let form = Form::new()
            .part("file", part)
            .text("key", key.to_string());
        let url = format!("{}{}", self.base_url, path);
        let mut req = self.client.post(&url).multipart(form);
        // Token refresh failed — proceed as unauthenticated
        if let Ok(key) = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| self.service_key.clone())) {
            if !key.is_empty() {
                req = req.header("X-EdgeBase-Service-Key", &key);
                req = req.header("Authorization", format!("Bearer {}", key));
            }
        }
        self.send(req).await
    }

    /// POST raw bytes (for multipart upload-part).
    pub async fn post_bytes(&self, path: &str, data: Vec<u8>, content_type: &str) -> Result<Value, Error> {
        let req = self.build_request(Method::POST, path)
            .header("Content-Type", content_type)
            .body(data);
        self.send(req).await
    }

    /// Download raw bytes.
    pub async fn download_raw(&self, path: &str) -> Result<Vec<u8>, Error> {
        let req = self.build_request(Method::GET, path);
        let resp = req.send().await?;
        let status = resp.status();
        if !status.is_success() {
            let msg = resp.text().await.unwrap_or_default();
            return Err(Error::Api { status: status.as_u16(), message: msg });
        }
        Ok(resp.bytes().await.map(|b| b.to_vec())?)
    }
}

File: packages/sdk/dart/packages/core/lib/src/http_client.dart
// HTTP client for EdgeBase API communication.
//
// Mirrors `@edge-base/sdk` HttpClient — handles auth headers, JSON
// serialization, error parsing, and automatic 401 retry with token refresh.

import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;
import 'package:http/http.dart' as http;
import 'errors.dart';
import 'token_manager.dart';
import 'context_manager.dart';
import 'http_client_platform_io.dart'
    if (dart.library.html) 'http_client_platform_web.dart' as platform;

class HttpClient {
  final String baseUrl;
  final String? serviceKey;
  final TokenManager? tokenManager;
  final ContextManager contextManager;
  final http.Client _client;
  final Duration _requestTimeout;
  String? _locale;

  factory HttpClient({
    required String baseUrl,
    TokenManager? tokenManager,
    required ContextManager contextManager,
    String? serviceKey,
    http.Client? client,
    Duration? requestTimeout,
  }) {
    final resolvedTimeout = requestTimeout ?? _resolveRequestTimeout();
    return HttpClient._internal(
      baseUrl: baseUrl,
      tokenManager: tokenManager,
      contextManager: contextManager,
      serviceKey: serviceKey,
      client: client ?? _createDefaultClient(resolvedTimeout),
      requestTimeout: resolvedTimeout,
    );
  }

  HttpClient._internal({
    required this.baseUrl,
    this.tokenManager,
    required this.contextManager,
    this.serviceKey,
    required http.Client client,
    required Duration requestTimeout,
  })  : _client = client,
        _requestTimeout = requestTimeout;

  static Duration _resolveRequestTimeout() {
    final raw = platform.platformEnvironmentValue('EDGEBASE_HTTP_TIMEOUT_MS');
    final milliseconds = raw == null ? null : int.tryParse(raw.trim());
    if (milliseconds != null && milliseconds > 0) {
      return Duration(milliseconds: milliseconds);
    }
    return const Duration(seconds: 30);
  }

  static http.Client _createDefaultClient(Duration timeout) =>
      platform.createDefaultHttpClient(timeout);

  /// Perform token refresh via HTTP.
  Future<TokenPair> _refreshToken(String refreshToken) async {
    final response = await _client
        .post(
          Uri.parse('$baseUrl/api/auth/refresh'),
          headers: {
            'Content-Type': 'application/json',
            'Connection': 'close',
          },
          body: jsonEncode({'refreshToken': refreshToken}),
        )
        .timeout(_requestTimeout);
    if (response.statusCode >= 200 && response.statusCode < 300) {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      return TokenPair(
        accessToken: json['accessToken'] as String,
        refreshToken: json['refreshToken'] as String,
      );
    }
    throw EdgeBaseError('Token refresh failed',
        statusCode: response.statusCode);
  }

  /// Headers with auth token and request metadata.
  Future<Map<String, String>> _buildHeaders({bool withAuth = true}) async {
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'Connection': 'close',
    };

    if (serviceKey != null) {
      // Service-key requests must use the dedicated header so client-facing
      // /api routes do not try to parse the key as a user JWT.
      headers['X-EdgeBase-Service-Key'] = serviceKey!;
    } else if (withAuth) {
      try {
        final token = await tokenManager?.getAccessToken(_refreshToken);
        if (token != null) {
          headers['Authorization'] = 'Bearer $token';
        }
      } catch (_) {
        // Token refresh failed — proceed as unauthenticated
      }
    }
    if (_locale != null && _locale!.isNotEmpty) {
      headers['Accept-Language'] = _locale!;
    }
    return headers;
  }

  /// Set locale for i18n/auth email language and Accept-Language headers.
  void setLocale(String? locale) {
    _locale = locale;
  }

  /// Get the currently configured locale override.
  String? getLocale() => _locale;

  /// Parse response, throw [EdgeBaseError] on failure.
  dynamic _handleResponse(http.Response response) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      if (response.body.isEmpty || response.statusCode == 204) return null;
      return jsonDecode(response.body);
    }
    // Try parse error JSON
    try {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      throw EdgeBaseError.fromJson(json, response.statusCode);
    } catch (e) {
      if (e is EdgeBaseError) rethrow;
      throw EdgeBaseError(
        'Request failed with status ${response.statusCode}',
        statusCode: response.statusCode,
      );
    }
  }

  /// Core request method with automatic 401 retry, 429 retry, and transport retry.
  Future<dynamic> _request(
    String method,
    String path, {
    Object? body,
    bool skipAuth = false,
    Map<String, String>? query,
  }) async {
    final uri = _buildUri(path, query);
    final encoded = body != null ? jsonEncode(body) : null;

    for (var attempt = 0; attempt <= 3; attempt++) {
      final headers = await _buildHeaders(withAuth: !skipAuth);

      http.Response response;
      try {
        response = await _send(method, uri, headers, encoded);
      } on TimeoutException {
        if (attempt < 2) {
          await Future.delayed(Duration(milliseconds: 50 * (attempt + 1)));
          continue;
        }
        throw EdgeBaseError(
          'Request timeout after ${_requestTimeout.inMilliseconds}ms',
        );
      } catch (e) {
        if (e is EdgeBaseError) rethrow;
        if (attempt < 2 && _isRetryableTransportError(e)) {
          await Future.delayed(Duration(milliseconds: 50 * (attempt + 1)));
          continue;
        }
        throw EdgeBaseError('Network error: $e');
      }

      // 429 retry with Retry-After
      if (response.statusCode == 429 && attempt < 3) {
        await Future.delayed(_parseRetryAfter(response, attempt));
        continue;
      }

      // 401 auto-retry: refresh token and retry once
      if (response.statusCode == 401 && !skipAuth && serviceKey == null && attempt == 0) {
        try {
          final newHeaders = await _buildHeaders(withAuth: true);
          response = await _send(method, uri, newHeaders, encoded);
          if (response.statusCode >= 200 && response.statusCode < 300) {
            return _handleResponse(response);
          }
        } catch (_) {
          // Retry failed, fall through to original response handling
        }
      }

      return _handleResponse(response);
    }

    throw EdgeBaseError('Request failed after maximum retries');
  }

  /// Build URI with optional query parameters.
  Uri _buildUri(String path, Map<String, String>? query) {
    final base = Uri.parse('$baseUrl/api$path');
    if (query != null && query.isNotEmpty) {
      return base.replace(queryParameters: {
        ...base.queryParameters,
        ...query,
      });
    }
    return base;
  }

  /// Send HTTP request.
  Future<http.Response> _send(
    String method,
    Uri uri,
    Map<String, String> headers,
    String? body,
  ) async {
    Future<http.Response> request() {
      switch (method) {
        case 'GET':
          return _client.get(uri, headers: headers);
        case 'POST':
          return _client.post(uri, headers: headers, body: body);
        case 'PATCH':
          return _client.patch(uri, headers: headers, body: body);
        case 'PUT':
          return _client.put(uri, headers: headers, body: body);
        case 'DELETE':
          return _client.delete(uri, headers: headers, body: body);
        case 'HEAD':
          return _client.head(uri, headers: headers);
        default:
          throw EdgeBaseError('Unsupported HTTP method: $method');
      }
    }

    return request().timeout(_requestTimeout);
  }

  // ─── Public API (auth required) ───

  Future<dynamic> get(String path, [Map<String, String>? query]) =>
      _request('GET', path, query: query);

  Future<dynamic> post(String path, [Object? body]) =>
      _request('POST', path, body: body);

  Future<dynamic> postWithQuery(
    String path,
    Object? body, [
    Map<String, String>? query,
  ]) =>
      _request('POST', path, body: body, query: query);

  Future<dynamic> patch(String path, Object? body) =>
      _request('PATCH', path, body: body);

  Future<dynamic> put(String path, Object? body) =>
      _request('PUT', path, body: body);

  Future<dynamic> putWithQuery(
    String path,
    Object? body, [
    Map<String, String>? query,
  ]) =>
      _request('PUT', path, body: body, query: query);

  Future<dynamic> delete(String path, [Object? body]) =>
      _request('DELETE', path, body: body);

  // ─── Public API (no auth, for signup/signin) ───

  Future<dynamic> postPublic(String path, [Object? body]) =>
      _request('POST', path, body: body, skipAuth: true);

  /// Get auth headers (for raw fetch calls, e.g. file uploads).
  Future<Map<String, String>> getAuthHeaders() => _buildHeaders(withAuth: true);

  /// Get raw [http.Response] for advanced use (e.g. file download).
  Future<http.Response> getRaw(String path) async {
    final headers = await _buildHeaders();
    return _client
        .get(
          Uri.parse('$baseUrl/api$path'),
          headers: headers,
        )
        .timeout(_requestTimeout);
  }

  /// POST raw bytes (for multipart upload-part).
  Future<dynamic> postRaw(String path, List<int> data,
      {String contentType = 'application/octet-stream'}) async {
    final headers = await _buildHeaders();
    headers['Content-Type'] = contentType;
    final response = await _client
        .post(
          Uri.parse('$baseUrl/api$path'),
          headers: headers,
          body: data,
        )
        .timeout(_requestTimeout);
    return _handleResponse(response);
  }

  /// Send multipart request (for file uploads).
  Future<dynamic> postMultipart(
    String path,
    http.MultipartRequest request,
  ) async {
    final headers = await _buildHeaders();
    headers.remove('Content-Type'); // Let multipart set its own
    request.headers.addAll(headers);
    final streamedResponse =
        await _client.send(request).timeout(_requestTimeout);
    final response = await http.Response.fromStream(streamedResponse);
    return _handleResponse(response);
  }

  /// HEAD request — returns true if resource exists (2xx).
  Future<bool> head(String path) async {
    final uri = Uri.parse('$baseUrl/api$path');
    final headers = await _buildHeaders();
    try {
      final response =
          await _client.head(uri, headers: headers).timeout(_requestTimeout);
      return response.statusCode >= 200 && response.statusCode < 300;
    } catch (_) {
      return false;
    }
  }

  static Duration _parseRetryAfter(http.Response response, int attempt) {
    final header = response.headers['retry-after'];
    var baseMs = 1000 * (1 << attempt);
    if (header != null) {
      final seconds = int.tryParse(header);
      if (seconds != null && seconds > 0) baseMs = seconds * 1000;
    }
    final random = math.Random();
    final jitter = (baseMs * 0.25 * random.nextDouble()).round();
    return Duration(milliseconds: (baseMs + jitter).clamp(0, 10000));
  }

  static bool _isRetryableTransportError(Object error) {
    final msg = error.toString().toLowerCase();
    return msg.contains('timeout') || msg.contains('socket') ||
        msg.contains('connection') || msg.contains('reset') ||
        msg.contains('refused') || msg.contains('network');
  }

  void close() => _client.close();
}
